library/core/src/any.rs

```diff
@@ -700,3 +769,201 @@ pub const fn type_name<T: ?Sized>() -> &'static str {
 pub const fn type_name_of_val<T: ?Sized>(_val: &T) -> &'static str {
     type_name::<T>()
 }
+
+///////////////////////////////////////////////////////////////////////////////
+// Provider trait
+///////////////////////////////////////////////////////////////////////////////
+
+/// Trait implemented by a type which can dynamically provide values based on type.
+#[unstable(feature = "provide_any", issue = "none")]
+pub trait Provider {
+    /// Object providers should implement this method to provide *all* values they are able to
+    /// provide using `req`.
+    #[unstable(feature = "provide_any", issue = "none")]
+    fn provide<'a>(&'a self, req: &mut Demand<'a>);
+}
+
+/// Request a value from the `Provider`.
+#[unstable(feature = "provide_any", issue = "none")]
+pub fn request_value<'a, T: 'static>(provider: &'a dyn Provider) -> Option<T> {
+    request_by_type_tag::<'a, tags::Value<T>>(provider)
+}
+
+/// Request a reference from the `Provider`.
+#[unstable(feature = "provide_any", issue = "none")]
+pub fn request_ref<'a, T: ?Sized + 'static>(provider: &'a dyn Provider) -> Option<&'a T> {
+    request_by_type_tag::<'a, tags::Ref<tags::MaybeSizedValue<T>>>(provider)
+}
+
+/// Request a specific value by tag from the `Provider`.
+fn request_by_type_tag<'a, I>(provider: &'a dyn Provider) -> Option<I::Reified>
+where
+    I: tags::Type<'a>,
+{
+    let mut tagged = TaggedOption::<'a, I>(None);
+    provider.provide(tagged.as_demand());
+    tagged.0
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Demand and its methods
+///////////////////////////////////////////////////////////////////////////////
+
+/// A helper object for providing objects by type.
+///
+/// An object provider provides values by calling this type's provide methods.
+#[allow(missing_debug_implementations)]
+#[unstable(feature = "provide_any", issue = "none")]
+// SAFETY: `TaggedOption::as_demand` relies on this precise definition.
+#[repr(transparent)]
+pub struct Demand<'a>(dyn Erased<'a> + 'a);
```

Comment:
I'm not sure I understand this suggestion. Are you suggesting that the end user use `MyType::provide(&Demand)` rather than `MyType::get_context()` (or whatever the end type names their getter method)? I'm not sure I understand the motivation in that case, using an explicit `Demand` to *get* the type seems strictly more complex than using a type-driven getter - there is an extra type to use (`Demand`) and extra steps which are seem to me to be boilerplate (setting up the `Demand` and then getting the value out). Compared to the getter, where the user just uses `let num_demanded: usize = my_type.get_context();`.

Reply:
Ah, yea, ignore this one. This was an aesthetic preference when I was reading through the docs explaining the feature that turned into an attempt to do crimes, not something I thought was seriously a good idea. I looked into making this work with unbound lifetimes but it ended up requiring a bunch of explicitly inferred generic parameters and definitely doesn't accomplish any of the goals I had (making it easier to teach).